src/state_space/mod.rs

```diff
@@ -131,15 +132,15 @@ where
     }
 
     pub async fn sync(self) -> Result<StateSpaceManager<T, N, P>, AMMError> {
-        let chain_tip = self.provider.get_block_number().await?;
-
         let mut futures = FuturesUnordered::new();
         let factories = self.factories.clone();
         for factory in factories {
             let provider = self.provider.clone();
             let filters = self.filters.clone();
             futures.push(tokio::spawn(async move {
-                let mut amms = factory.discover(chain_tip, provider.clone()).await?;
+                let mut amms = factory
+                    .discover(BlockId::latest(), provider.clone())
```

Comment:
We need to get fetch `let chain_tip = self.provider.get_block_number().await?;` and pass it in for all factories. Otherwise, it may possible to discovery and sync at different block heights across factories.